vault/identity_store_oidc_provider.go

```diff
@@ -1504,20 +1588,473 @@ func authResponse(code, state, errorCode, errorDescription string) (*logical.Res
+		return tokenResponse(nil, ErrTokenInvalidGrant, "redirect_uri does not match the redirect_uri used in the authorization request")
+	}
+
+	// Get the entity associated with the initial authorization request
+	entity, err := i.MemDBEntityByID(authCodeEntry.entityID, true)
+	if err != nil {
+		return tokenResponse(nil, ErrTokenServerError, err.Error())
+	}
+	if entity == nil {
+		return tokenResponse(nil, ErrTokenInvalidRequest, "identity entity associated with request not found")
+	}
+
+	// Validate that the entity is a member of the client's assignments
+	isMember, err := i.entityHasAssignment(ctx, req.Storage, entity.GetID(), client.Assignments)
+	if err != nil {
+		return tokenResponse(nil, ErrTokenServerError, err.Error())
+	}
+	if !isMember {
+		return tokenResponse(nil, ErrTokenInvalidRequest, "identity entity not authorized by client assignment")
+	}
+
+	// The access token is a Vault batch token with a policy that only
+	// provides access to the issuing provider's userinfo endpoint.
+	accessTokenIssuedAt := time.Now()
+	accessTokenExpiry := accessTokenIssuedAt.Add(client.AccessTokenTTL)
+	accessToken := &logical.TokenEntry{
+		Type:         logical.TokenTypeBatch,
+		NamespaceID:  ns.ID,
+		Path:         req.Path,
+		TTL:          client.AccessTokenTTL,
+		CreationTime: accessTokenIssuedAt.Unix(),
+		EntityID:     entity.ID,
+
+		// TODO: Move this to InternalMeta once support is checked in
+		Meta: map[string]string{
+			accessTokenClientIDMeta: client.ClientID,
+			accessTokenScopesMeta:   strings.Join(authCodeEntry.scopes, scopesDelimiter),
+		},
+
+		// TODO: Add the InlinePolicy once support is checked in
+		//InlinePolicy: fmt.Sprintf(`
+		//	path "identity/oidc/provider/%s/userinfo" {
+		//		capabilities = ["read", "update"]
+		//	}
+		//`, name),
+	}
+	err = i.tokenStorer.TokenStore().create(ctx, accessToken)
+	if err != nil {
+		return tokenResponse(nil, ErrTokenServerError, err.Error())
+	}
+
+	// Compute the access token hash claim (at_hash)
+	atHash, err := computeHashClaim(key.Algorithm, accessToken.ID)
+	if err != nil {
+		return tokenResponse(nil, ErrTokenServerError, err.Error())
+	}
+
+	// Compute the authorization code hash claim (c_hash)
+	cHash, err := computeHashClaim(key.Algorithm, code)
+	if err != nil {
+		return tokenResponse(nil, ErrTokenServerError, err.Error())
+	}
+
+	// Set the ID token claims
+	idTokenIssuedAt := time.Now()
+	idTokenExpiry := idTokenIssuedAt.Add(client.IDTokenTTL)
+	idToken := idToken{
+		Namespace:       ns.ID,
+		Issuer:          provider.effectiveIssuer,
+		Subject:         authCodeEntry.entityID,
+		Audience:        authCodeEntry.clientID,
+		Nonce:           authCodeEntry.nonce,
+		Expiry:          idTokenExpiry.Unix(),
+		IssuedAt:        idTokenIssuedAt.Unix(),
+		AccessTokenHash: atHash,
+		CodeHash:        cHash,
+	}
+
+	// Add the auth_time claim if it's not the zero time instant
+	if !authCodeEntry.authTime.IsZero() {
+		idToken.AuthTime = authCodeEntry.authTime.Unix()
```

Comment:
If I am understanding `auth_time` correctly, this claim is optional for us. Is that correct?

Reply:
That's correct. The `auth_time` claim is optional _unless_ the `max_age` parameter was present in the authentication request.

> When a max_age request is made or when auth_time is requested as an Essential Claim, then this Claim is REQUIRED; otherwise, its inclusion is OPTIONAL.